lock-api/src/test/java/com/palantir/lock/client/LeaderElectionReportingTimelockServiceTest.java

```diff
@@ -101,4 +115,207 @@ public void leaderIsInitialisedForAllMethods() {
         verify(mockedRegistry).timer(any());
         verify(mockedTimer).update(anyLong(), any());
     }
+
+    @Test
+    public void noLeaderElectionDurationBeforeLeaderElection() {
+        assertThat(timelockService.calculateLastLeaderElectionDuration()).isNotPresent();
+
+        timelockService.startTransactions(startTransactionsRequest);
+        assertThat(timelockService.calculateLastLeaderElectionDuration()).isNotPresent();
+
+        timelockService.getCommitTimestamps(commitTimestampsRequest);
+        assertThat(timelockService.calculateLastLeaderElectionDuration()).isNotPresent();
+    }
+
+    /**
+     * [ A ]
+     *       [ A ]
+     *             [ B ]
+     *       <=========>
+     */
+    @Test
+    public void detectLeaderElectionWithFreshLeader() {
+        executeCalls(
+                ImmutableSingleCall.of(0L, 1L, LEADER_1),
+                ImmutableSingleCall.of(3L, 6L, LEADER_1),
+                ImmutableSingleCall.of(10L, 15L, LEADER_2));
+        assertExpectedDuration(Instant.ofEpochMilli(3L), Instant.ofEpochMilli(15L));
```

Comment:
hm - given that the second call returns at 6L, why is the expected duration not 6L -> 15L? I realise that I'm probably misunderstanding your massive and lengthy explanation above, but when laid out this way, it just seems weird

Reply:
I think the theory here is that we want to have an upper bound, and what could have happened is (where the epsilon_i are small values ordered parallel to i)

T = 3, A makes request
T = 3 + epsilon, the leadership check on timelock passes, and timelock starts running the actual code that was protected by a leader election service
T = 3 + epsilon + mu, timelock leader elects
T = 6, the response finally comes back (note: this looks spooky but is allowed)

In general, if you have a request running 3 to 6 and this is the last request you know about that had a leader A, you know that the election didn't begin before 3, but you don't know that the election hadn't begun at any point after 3.

Reply:
When you put it that way, it's obvious - the upper bound here being the last successfully sent request. Thanks!